src/librustc_lint/types.rs

```diff
@@ -832,9 +833,19 @@ impl<'a, 'tcx> ImproperCTypesVisitor<'a, 'tcx> {
             },
 
             ty::RawPtr(ty::TypeAndMut { ty, .. }) |
-            ty::Ref(_, ty, _) => self.check_type_for_ffi(cache, ty),
+            ty::Ref(_, ty, _) => self.check_type_for_ffi(cache, ty, false),
+
+            ty::Array(..) if top_level => {
+                // C doesn't really support passing arrays by value.
+                // The only way to pass an array by value is through a struct.
+                FfiUnsafe {
+                    ty,
+                    reason: "passing raw arrays by value is not FFI-safe",
+                    help: Some("consider passing a pointer to the array"),
+                }
+            }
```

Comment:
I think it's really easy to poison the cache, so maybe add a method that wraps `check_type_for_ffi` (say, `check_top_level_type_for_ffi`?) and checks for `ty::Array` before calling `check_type_for_ffi`?

Reply:
Hmm, the point of the cache is to prevent infinite recursion. 
I'm not sure I'm seeing how is it any different than other checks.

Reply:
`top_level` isn't in the hash key, so whichever comes first will stay in the cache and affect any other occurrences of the same type.

Also, overall the diff would be simpler if the top-level check was separate, and `check_type_for_ffi` didn't have an extra parameter.

Reply:
Updated the name of the inner one, otherwise calling `check_top_level_type_for_ffi` would've been misleading, because it checks both top level and lower levels.